onnx/defs/generator/defs.cc

```diff
@@ -523,4 +527,185 @@ ONNX_OPERATOR_SET_SCHEMA(
+up to `limit` (exclusive).)DOC";
+
+template <typename T>
+inline int compute_output_dim_for_range(
+    const TensorProto* start,
+    const TensorProto* limit,
+    const TensorProto* delta) {
+  const auto& start_data = ParseData<T>(start);
+  const auto& limit_data = ParseData<T>(limit);
+  const auto& delta_data =
+      delta != nullptr ? ParseData<T>(delta) : std::vector<T>(1, 1);
+
+  if (start_data.size() != 1 || limit_data.size() != 1 ||
+      delta_data.size() != 1) {
+    fail_shape_inference(
+        "Input to 'Range' op should be scalars (Tensor with only one element)");
+  }
+
+  int n = static_cast<int>(
+      ceil((1.0 * (limit_data[0] - start_data[0])) / delta_data[0]));
+  if (n < 0)
+    n = 0;
+
+  return n;
+}
+
+const std::vector<NodeProto> build_nodes_range_op() {
+    // body for 'Loop node'
+    GraphProto loop_sub_graph;
+    loop_sub_graph.set_name("loop body attribute");
+
+    // 'Loop' node 'body' attribute's graph inputs
+    auto* input_value_info_proto_0 = loop_sub_graph.add_input();
+    input_value_info_proto_0->set_name("i");
+
+    auto* input_value_info_proto_1 = loop_sub_graph.add_input();
+    input_value_info_proto_1->set_name("cond");
+
+    auto* input_value_info_proto_2 = loop_sub_graph.add_input();
+    input_value_info_proto_2->set_name("prev");
+
+    // 'Loop' node 'body' attribute's graph nodes
+    auto* node_proto_0 = loop_sub_graph.add_node();
+    node_proto_0->set_op_type("Identity");
+    node_proto_0->add_input();
+    node_proto_0->set_input(0, "cond");
+    node_proto_0->add_output();
+    node_proto_0->set_output(0, "cond_out");
+
+    auto* node_proto_1 = loop_sub_graph.add_node();
+    node_proto_1->set_op_type("Add");
+    node_proto_1->add_input();
+    node_proto_1->set_input(0, "prev");
+    node_proto_1->add_input();    
+    node_proto_1->set_input(1, "delta");
+    node_proto_1->add_output();
+    node_proto_1->set_output(0, "current");
+
+    auto* node_proto_2 = loop_sub_graph.add_node();
+    node_proto_2->set_op_type("Identity");
+    node_proto_2->add_input();
+    node_proto_2->set_input(0, "prev");
+    node_proto_2->add_output();
+    node_proto_2->set_output(0, "range");
+
+    // 'Loop' node 'body' attribute's graph inputs
+    auto* output_value_info_proto_0 = loop_sub_graph.add_output();
+    output_value_info_proto_0->set_name("cond_out");
+
+    auto* output_value_info_proto_1 = loop_sub_graph.add_output();
+    output_value_info_proto_1->set_name("current");
+
+    auto* output_value_info_proto_2 = loop_sub_graph.add_output();
+    output_value_info_proto_2->set_name("range");
+
+    return FunctionBodyHelper::BuildNodes({
+      // nodes: {outputs, op, inputs, attributes}
+      {{"sub_result"}, "Sub", {"limit", "start"}},
+      {{"delta_cast"}, "Cast", {"delta"}, {{"to", static_cast<int64_t>(1)}}},
+      {{"div_result"}, "Div", {"sub_result", "delta_cast"}},
+      {{"ceil_result"}, "Ceil", {"div_result"}},
```

Comment:
If `Ceil` is used, could the `output` contain elements larger than `limit`?

Reply:
I don't think so. 

The number of elements for range is computed as: 

Max(ceil((limit - start) / delta), 0)

like here:

https://github.com/microsoft/onnxruntime/blob/bf6a9f9c27e14cf334e6cf30e8a91c39f2456728/onnxruntime/contrib_ops/cpu/range.cc#L38

Reply:
I just some manual calculation. You're right. Thanks.